telemetry/appblock.go

```diff
@@ -0,0 +1,105 @@
+package telemetry
+
+import (
+	"runtime"
+	"strconv"
+	"time"
+
+	abci "github.com/cometbft/cometbft/abci/types"
+	"github.com/cosmos/cosmos-sdk/baseapp"
+	"github.com/cosmos/cosmos-sdk/telemetry"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/mempool"
+	"github.com/hashicorp/go-metrics"
+)
+
+type BlockTelemetry struct {
+	startTime            time.Time
+	initialBlockMemAlloc uint64
+	blockTxsSize         int
+}
+
+// InitBlockMetrics initializes the block metrics for the application
+// by initializing a BlockTelemetry struct and setting the prepare proposal,
+// begin blocker, and end blocker handlers with the metrics.
+func InitBlockMetrics(app *baseapp.BaseApp) {
+	blockMetrics := &BlockTelemetry{
+		startTime:            time.Now(),
+		initialBlockMemAlloc: 0,
+		blockTxsSize:         0,
+	}
+
+	app.SetPrepareProposal(initPrepareProposalHandlerWithMetrics(app, blockMetrics))
+	app.SetPrecommiter(initPreCommitterWithMetrics(blockMetrics))
+}
+
+// initPrepareProposalHandlerWithMetrics initializes the prepare proposal handler
+// with the block metrics.
+// It gathers the block txs size to emit them as a gauge metric.
+// It follows the NewBaseApp implementation of the PrepareProposalHandler by
+// setting the mempool to NoOpMempool and returning the prepare proposal handler.
+// that is then wrapped in a closure to gather the block txs size before handing
+// it off to the prepare proposal handler.
+func initPrepareProposalHandlerWithMetrics(
+	app *baseapp.BaseApp,
+	blockMetrics *BlockTelemetry,
+) sdk.PrepareProposalHandler {
+	// Create a NoOpMempool for the application and get the default prepare proposal
+	// handler as per NewBaseApp implementation.
+	// See https://github.com/cosmos/cosmos-sdk/blob/v0.50.4/baseapp/baseapp.go#L221
+	app.SetMempool(mempool.NoOpMempool{})
```

Comment:
It looks like the default implementation only sets the no-op mempool if it's nill: https://github.com/cosmos/cosmos-sdk/blob/v0.50.4/baseapp/baseapp.go#L214.

Reply:
It is already nil in our case and we need to pass-in a mempool to `NewDefaultProposalHandler`.

If we choose to have another mempool in the future, this is where is should be changed.